core/commands/files/files.go

```diff
@@ -0,0 +1,555 @@
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		dashp, _, _ := req.Option("parents").Bool()
+		dirtomake := req.Arguments()[0]
+
+		if dirtomake[0] != '/' {
+			res.SetError(errors.New("paths must be absolute"), cmds.ErrNormal)
+			return
+		}
+
+		err = mfs.Mkdir(n.FilesRoot, dirtomake, dashp)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+var FilesRmCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline:          "remove a file",
+		ShortDescription: ``,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, true, "file to remove"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("r", "recursive", "recursively remove directories"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		nd, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[0]
+		dir, name := gopath.Split(path)
+		parent, err := mfs.Lookup(nd.FilesRoot, dir)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		pdir, ok := parent.(*mfs.Directory)
+		if !ok {
+			res.SetError(fmt.Errorf("no such file or directory: %s", path), cmds.ErrNormal)
+			return
+		}
+
+		childi, err := pdir.Child(name)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		dashr, _, _ := req.Option("r").Bool()
+
+		switch childi.(type) {
+		case *mfs.Directory:
+			if dashr {
+				err := pdir.Unlink(name)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+			} else {
+				res.SetError(fmt.Errorf("%s is a directory, use -r to remove directories", path), cmds.ErrNormal)
+				return
+			}
+		default:
+			err := pdir.Unlink(name)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		}
+	},
+}
+
+func getFileHandle(r *mfs.Root, path string, create bool) (*mfs.File, error) {
+
+	target, err := mfs.Lookup(r, path)
+	switch err {
+	case nil:
+		fi, ok := target.(*mfs.File)
+		if !ok {
+			return nil, fmt.Errorf("%s was not a file", path)
+		}
+		return fi, nil
+
+	case os.ErrNotExist:
+		if !create {
+			return nil, err
+		}
+
+		// if create is specified and the file doesnt exist, we create the file
+		dirname, fname := gopath.Split(path)
+		pdiri, err := mfs.Lookup(r, dirname)
+		if err != nil {
+			log.Error("lookupfail ", dirname)
+			return nil, err
+		}
+		pdir, ok := pdiri.(*mfs.Directory)
+		if !ok {
+			return nil, fmt.Errorf("%s was not a directory", dirname)
+		}
+
+		nd := &dag.Node{Data: ft.FilePBData(nil, 0)}
+		err = pdir.AddChild(fname, nd)
+		if err != nil {
+			return nil, err
+		}
+
+		fsn, err := pdir.Child(fname)
+		if err != nil {
+			return nil, err
+		}
+
+		// can unsafely cast, if it fails, that means programmer error
+		return fsn.(*mfs.File), nil
```

Comment:
yeah, i guess if someone were to rm the file right after the `AddChild` call and mkdir with the same name before the Child call. I'll put a check in.